File: scripts/verify-successor-materialization-wire.mjs
/** Actual CTO Python inspection/publication -> gateway receipt acceptance.
 * node --import tsx scripts/verify-successor-materialization-wire.mjs MATERIALIZER_DIRECTORY [PYTHON]
 * Synthetic in-memory source only; no AWS calls or model execution.
 */
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { verifyMaterializationReceipt } from '../src/server/materialized-catalog-pin.ts';

assert.ok(process.argv[2], 'MATERIALIZER_DIRECTORY is required');
const canonicalSource="const canonical=(v:any):string=>v===null||typeof v!=='object'?JSON.stringify(v):Array.isArray(v)?'['+v.map(canonical).join(',')+']':'{'+Object.keys(v).sort().map(k=>JSON.stringify(k)+':'+canonical(v[k])).join(',')+'}';";
for(const path of ['../src/server/graph-catalog-controller.ts','../src/server/relationship-historical-read.ts']){
  assert.ok(readFileSync(new URL(path,import.meta.url),'utf8').includes(canonicalSource),`gateway canonical producer changed: ${path}`);
}
// The two gateway producers are transitively import-bound to runtime env validation.
// Assert their exact serializer source, then execute that source-equivalent function
// to emit the bytes consumed by the Python source-owner capture below.
const gatewayCanonical=value=>value===null||typeof value!=='object'?JSON.stringify(value):Array.isArray(value)?`[${value.map(gatewayCanonical).join(',')}]`:`{${Object.keys(value).sort().map(key=>`${JSON.stringify(key)}:${gatewayCanonical(value[key])}`).join(',')}}`;
const python=process.argv[3] ?? 'python', producerCanonical=gatewayCanonical,
  grantCanonical=gatewayCanonical,sha=value=>createHash('sha256').update(value).digest('hex');
const runId='run_'+sha('gateway-producer-run'), proposalKey=sha('gateway-producer-proposal'),
  manifestSha=sha('gateway-producer-manifest'), callerHash=sha('gateway-producer-caller'), policyHash=sha('gateway-producer-policy'),
  documentVersion='docv_'+sha('gateway-producer-document');
const proposal={key:proposalKey,run:{ref_version:'neptune-trial-active-run-ref-v1',run_id:runId,purpose:'company_graph_backfill',scope:'finance',run_version:'synthetic-v1',manifest_sha256:manifestSha},manifest:{manifest_sha256:manifestSha,documents:[{document_version_id:documentVersion,source_version:sha('gateway-producer-source'),source_path_hash:sha('gateway-producer-path')}]}},
  unsignedAdmission={allowed:true,key:proposalKey,run_id:runId,manifest_sha256:manifestSha,max_documents:1,policy_sha256:policyHash},
  admission={...unsignedAdmission,decision_sha256:sha(producerCanonical(unsignedAdmission))};
const proposalKeyPath=`graph-trial/20260908/catalog-cohorts/cfo-catalog-publish-20260909-live/server/proposals/${proposalKey}.json`,
  admissionKeyPath=`graph-trial/20260908/catalog-cohorts/cfo-catalog-publish-20260909-live/server/admissions/${runId}.json`,
  proposalRaw=Buffer.from(producerCanonical(proposal)), admissionRaw=Buffer.from(producerCanonical(admission)),
  grantKeyPath=`graph-trial/20260908/relationship-publications/cfo/cfo-catalog-publish-20260909-live/cfo-relationship-worker/runs/${runId}.json`,
  grant={schema:'relationship-publication-grant-v1',cohort_id:'cfo-catalog-publish-20260909-live',producer_id:'cfo-relationship-worker',caller_hash:callerHash,run:proposal.run,
    admission:{key:admissionKeyPath,version_id:'admission-v1',sha256:sha(admissionRaw)},proposal:{key:proposalKeyPath,version_id:'proposal-v1',sha256:sha(proposalRaw)},artifact_ref:{},approved_artifacts:[],issued_under_policy_version:'policy-v1'},
  grantRaw=Buffer.from(grantCanonical(grant));
const producerFixture={config:{schema:'cfo-completed-publication-capture-config-v1',source_cohort_id:'cfo-catalog-publish-20260909-live',producer_id:'cfo-relationship-worker',expected_caller_hash:callerHash,expected_admission_policy_sha256:policyHash},objects:[
  {key:grantKeyPath,version_id:'grant-v1',raw_base64:grantRaw.toString('base64')},
  {key:proposalKeyPath,version_id:'proposal-v1',raw_base64:proposalRaw.toString('base64')},
  {key:admissionKeyPath,version_id:'admission-v1',raw_base64:admissionRaw.toString('base64')},
]};
const producerRaw=execFileSync(python,[fileURLToPath(new URL('./verify-successor-materialization-wire.py',import.meta.url)),process.argv[2],'--verify-gateway-producer-fixture'],{input:JSON.stringify(producerFixture),encoding:'utf8',maxBuffer:256*1024});
assert.deepEqual(JSON.parse(producerRaw),{status:'passed',captured_items:1});
const raw = execFileSync(python, [fileURLToPath(new URL('./verify-successor-materialization-wire.py', import.meta.url)), process.argv[2]], {encoding:'utf8', maxBuffer:256*1024});
const fixture = JSON.parse(raw), body = Buffer.from(fixture.receipt_base64, 'base64');
assert.equal(fixture.counts.eligible_rows, 1);
assert.equal(fixture.counts.excluded.unknown_extraction_quarantine, 2);
assert.equal(fixture.counts.excluded.already_published_source, 1);
assert.ok(fixture.completed_publication_exclusion_manifest);
assert.equal(fixture.completed_publication_exclusion_manifest.count, 1);
assert.ok(body.length <= 64*1024);
const verified = await verifyMaterializationReceipt(fixture.context, async request => {
  assert.equal(request.key, fixture.context.materialization.materialization_receipt_key);
  return {status:200, headers:new Headers({'x-amz-version-id':fixture.receipt_version}), body};
}, AbortSignal.timeout(5000));
assert.equal(verified.catalogVersionId, fixture.context.materialization.catalog_version_id);
console.log(JSON.stringify({schema:'successor-materialization-wire-check-v1', status:'passed', synthetic:true, aws_calls:0, eligible_rows:1, excluded_unknown_rows:2, excluded_completed_rows:1, receipt_bytes:body.length, gateway_producer_serialization:'passed'}));

File: scripts/verify-successor-materialization-wire.py
"""Produce a synthetic receipt using the actual CTO source-owned publisher.

Usage: python scripts/verify-successor-materialization-wire.py MATERIALIZER_DIRECTORY
Uses an in-memory S3 fixture. No AWS or document source access occurs.
"""
import base64
import hashlib
import json
import sys
from pathlib import Path

if len(sys.argv) < 2:
    raise SystemExit('MATERIALIZER_DIRECTORY is required')

sys.path.insert(0, str(Path(sys.argv[1]).resolve()))


def verify_gateway_producer_fixture():
    """Exercise capture's raw-byte check with objects serialized by gateway code.

    The Node caller supplies only synthetic proposal, admission, and publication
    grant bytes, each produced by the gateway's own canonical serializer.  This
    keeps this compatibility regression free of AWS and source-document input.
    """
    import capture_completed_publication_exclusions as capture

    payload = json.load(sys.stdin)
    if set(payload) != {'config', 'objects'} or not isinstance(payload['objects'], list):
        raise SystemExit('invalid producer fixture')

    class Body:
        def __init__(self, data): self.data = data
        def iter_chunks(self, chunk_size=65536):
            for offset in range(0, len(self.data), chunk_size):
                yield self.data[offset:offset + chunk_size]
        def close(self): pass

    class S3:
        def __init__(self, records): self.records = records
        def list_objects_v2(self, **kwargs):
            prefix = kwargs['Prefix']
            return {'IsTruncated': False, 'Contents': [
                {'Key': key} for key in self.records if key.startswith(prefix) and '/runs/' in key
            ]}
        def get_object(self, **kwargs):
            record = self.records[kwargs['Key']]
            if kwargs.get('VersionId') not in (None, record['version_id']):
                raise AssertionError('pinned version changed')
            return {'VersionId': record['version_id'], 'Body': Body(record['raw'])}

    records = {}
    for item in payload['objects']:
        if set(item) != {'key', 'version_id', 'raw_base64'}:
            raise SystemExit('invalid producer object')
        records[item['key']] = {'version_id': item['version_id'],
                                'raw': base64.b64decode(item['raw_base64'], validate=True)}
    captured = capture.capture_from_s3(S3(records), 'synthetic-bucket', payload['config'])
    print(json.dumps({'status': 'passed', 'captured_items': len(captured['items'])}, separators=(',', ':')))


if len(sys.argv) == 3 and sys.argv[2] == '--verify-gateway-producer-fixture':
    verify_gateway_producer_fixture()
    raise SystemExit(0)

import test_source_bound_publish as fixture

worker = fixture.source_bound
rows = [fixture.row('unknown/one.json'), fixture.row('unknown/two.json'),
        fixture.row('already-published.json', sha256='3' * 64, enriched_sha256='3' * 64),
        fixture.row('eligible.json', sha256='4' * 64, enriched_sha256='4' * 64)]
exclusions = []
for index, row in enumerate(rows[:2]):
    marker = str(index + 1) * 64
    exclusions.append({'proposal_key': marker, 'run_id': 'run_' + marker,
                       'operation_id': 'subop_' + marker, 'source_document_version': 'docv_' + marker,
                       'catalog_source_sha256': row['sha256'],
                       'source_path_sha256': hashlib.sha256(row['path'].encode()).hexdigest()})
completed_digest = hashlib.sha256(b'synthetic-old-publication').hexdigest()
completed_manifest = {'schema': 'cfo-completed-publication-exclusion-manifest-v1',
                      'source_cohort_id': 'cfo-catalog-publish-20260909-live',
                      'producer_id': 'cfo-relationship-worker',
                      'items': [{'proposal_key': completed_digest, 'run_id': 'run_' + completed_digest,
                                 'source_document_version': 'docv_' + completed_digest,
                                 'catalog_source_sha256': rows[2]['sha256'],
                                 'source_path_sha256': hashlib.sha256(rows[2]['path'].encode()).hexdigest(),
                                 'publication_grant': {'key': 'graph-trial/synthetic/grant', 'version_id': 'grant-v1', 'sha256': completed_digest},
                                 'proposal': {'key': 'graph-trial/synthetic/proposal', 'version_id': 'proposal-v1', 'sha256': completed_digest},
                                 'admission': {'key': 'graph-trial/synthetic/admission', 'version_id': 'admission-v1', 'sha256': completed_digest}}]}
completed_raw = worker.materialize.completed_manifest.canonical(completed_manifest)
completed_sha = hashlib.sha256(completed_raw).hexdigest()
completed_pointer = worker.materialize.completed_manifest.pointer(
    completed_manifest, worker.materialize.completed_manifest.PREFIX + completed_sha + '.json', 'completed-v1')
request = {'operation': 'inspect', 'source_version_id': 'source-v1',
           'cohort_id': 'cfo-catalog-successor-unknown-20260912', 'policy_sha256': 'b' * 64,
           'source_prefixes': [], 'source_scope': 'all_cfo_source_documents',
           'quarantine_exclusions': exclusions,
           'completed_publication_exclusion_manifest': completed_pointer}
authority = [{key: value for key, value in request.items() if key != 'operation'} |
             {'allow_publish': True, 'expires_at': '2099-01-01T00:00:00Z'}]
source = fixture.raw(*rows)
inspect_s3 = fixture.S3(source)
inspect_s3.objects[completed_pointer['key']] = {'data': completed_raw, 'version': completed_pointer['version_id']}
inspected = worker.materialize.materialize(request, inspect_s3, authority, fixture.NOW)
s3 = fixture.S3(source)
s3.objects[completed_pointer['key']] = {'data': completed_raw, 'version': completed_pointer['version_id']}
result = worker.publish_source_bound({'operation': 'publish_source_bound', 'inspect_receipt': inspected,
                                     'inspect_receipt_sha256': worker.publish_inspected.digest(worker.publish_inspected.encode(inspected))},
                                    s3, authority, fixture.NOW)
stored = s3.objects[result['receipt_key']]
context = {key: result[key] for key in ('cohort_id', 'policy_sha256', 'source_scope', 'catalog_key', 'catalog_source_sha256')}
context['source_prefixes'] = []
context['materialization'] = {'catalog_content_sha256': result['catalog_content_sha256'],
                              'source_catalog_version_id': result['source_version_id'],
                              'materialization_receipt_key': result['receipt_key'],
                              'materialization_receipt_sha256': result['receipt_sha256'],
                              'catalog_version_id': result['catalog_version_id']}
print(json.dumps({'context': context, 'receipt_base64': base64.b64encode(stored['data']).decode(),
                  'receipt_version': stored['version'], 'counts': result['counts'],
                  'completed_publication_exclusion_manifest': result['completed_publication_exclusion_manifest']}))
